src/android/java/io/jxcore/node/BtConnectorHelper.java

```diff
@@ -192,25 +194,32 @@ public String  MakeGUID() {
     public void BeginConnectPeer(String toPeerId, ConnectStatusCallback connectStatusCallback) {
         ConnectStatusCallback tmpCallback = connectStatusCallback;
 
+        // needed since mBtToRequestSocket can be disconnected from other thread
+        BtToRequestSocket tmp = mBtToRequestSocket;
+
         if(tmpCallback == null) {
             //nothing we should do, since we can not update progress
             print_debug("BeginConnectPeer callback is NULL !!!!!!");
-        }else if(mBtToRequestSocket != null) {
-            tmpCallback.ConnectionStatusUpdate("Already connected to " + mBtToRequestSocket.GetPeerId(),-1);
+        }else if(tmp != null) {
+            tmpCallback.ConnectionStatusUpdate("Already connected to " + tmp.GetPeerId(),-1);
         }else {
             ServiceItem selectedDevice = null;
             if (lastAvailableList != null) {
-                for (int i = 0; i < lastAvailableList.size(); i++) {
-                    if (lastAvailableList.get(i).peerId.contentEquals(toPeerId)) {
-                        selectedDevice = lastAvailableList.get(i);
-                        break;
+                synchronized (lastAvailableList) {
+                    for (int i = 0; i < lastAvailableList.size(); i++) {
+                        if (lastAvailableList.get(i).peerId.contentEquals(toPeerId)) {
+                            selectedDevice = lastAvailableList.get(i);
+                            break;
+                        }
                     }
                 }
             }
 
             if (selectedDevice != null) {
-                if (mBTConnector != null) {
-                    BTConnector.TryConnectReturnValues retVal = mBTConnector.TryConnect(selectedDevice);
+                BTConnector tmpConn = mBTConnector;
+                if (tmpConn != null) {
+                    // TryConnect is synchronized and thus thread safe call
+                    BTConnector.TryConnectReturnValues retVal = tmpConn.TryConnect(selectedDevice);
```

Comment:
@DrJukka I'm guessing line 225 below isn't correct since it presumes we can only connect to one peer at a time?